src/mtev_http.c

```diff
@@ -1197,20 +1221,57 @@ mtev_http_session_req_consume_chunked(mtev_http_session_ctx *ctx,
     }
     if(rlen <= 0) {
       mtevL(http_debug, " ... mtev_http_session_req_consume = -1 (error)\n");
-      return -1;
+      return -2;
     }
     in->size += rlen;
+    if (ctx->req.payload_chunked == mtev_false) {
+      next_chunk += rlen;
+    }
   }
 
  successful_chunk_size:
-  if(in->size == 0) {
-    tofree = in;
-    ctx->req.first_input = in = in->next;
-    if(ctx->req.last_input == tofree) ctx->req.last_input = in;
-    tofree->next = NULL;
-    RELEASE_BCHAIN(tofree);
-  }
+  {
+    if (next_chunk > 0) {
+      mtevL(http_debug, " ... have chunk (%d)\n", next_chunk);
+      struct bchain *data = ALLOC_BCHAIN(next_chunk);
+      data->compression = compression_type;
+      if (ctx->req.user_data_last != NULL) {
+        ctx->req.user_data_last->next = data;
+      }
+      ctx->req.user_data_last = data;
+      if (ctx->req.user_data == NULL) {
+        ctx->req.user_data = data;
+      }
+      memcpy(data->buff + data->size, in->buff + in->start, MIN(in->size, next_chunk));
+      data->size = MIN(in->size, next_chunk);
+      in->start += data->size;
+      in->size -= data->size;
+      ctx->req.user_data_last = data;
+      if (ctx->req.payload_chunked) {
+        /* there must be a \r\n at the end of this block */
+        str_in_f = strnstrn("\r\n", 2, in->buff + in->start, in->size);
+        if(str_in_f != (in->buff + in->start)) {
```

Comment:
I _think_ this can just be a `if (in->size < 2 || strcmp(in->buff + in->start, "\r\n") != 0) {`? `in->start` should now point to the end of the chunk, so it seems a little more correct, to me... more robust against possibly storing extra data in the `in` chunk.